ksqldb-engine/src/main/java/io/confluent/ksql/util/SharedKafkaStreamsRuntimeImpl.java

```diff
@@ -0,0 +1,225 @@
+
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import io.confluent.ksql.name.SourceName;
+import io.confluent.ksql.query.KafkaStreamsBuilder;
+import io.confluent.ksql.query.QueryError;
+import io.confluent.ksql.query.QueryErrorClassifier;
+import io.confluent.ksql.query.QueryId;
+import io.confluent.ksql.rest.entity.StreamsTaskMetadata;
+import java.time.Duration;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+import org.apache.kafka.streams.KafkaStreams;
+import org.apache.kafka.streams.LagInfo;
+import org.apache.kafka.streams.StreamsMetadata;
+import org.apache.kafka.streams.errors.StreamsException;
+import org.apache.kafka.streams.errors.StreamsUncaughtExceptionHandler;
+import org.apache.kafka.streams.processor.internals.namedtopology.KafkaStreamsNamedTopologyWrapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SharedKafkaStreamsRuntimeImpl implements SharedKafkaStreamsRuntime {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SharedKafkaStreamsRuntimeImpl.class);
+
+  private final KafkaStreamsBuilder kafkaStreamsBuilder;
+  private QueryErrorClassifier errorClassifier ;
+  private KafkaStreamsNamedTopologyWrapper kafkaStreams;
+  private ImmutableMap<String, Object> streamsProperties;
+  private final QueryMetadataImpl.TimeBoundedQueue queryErrors;
+  private final Map<String, PersistentQueriesInSharedRuntimesImpl> metadata;
+  public final Map<QueryId, Set<SourceName>> sources;
+
+  public SharedKafkaStreamsRuntimeImpl(final KafkaStreamsBuilder kafkaStreamsBuilder,
+                                       final int maxQueryErrorsQueueSize,
+                                       final Map<String, Object> streamsProperties) {
+    this.kafkaStreamsBuilder = kafkaStreamsBuilder;
+    kafkaStreams = kafkaStreamsBuilder.build(streamsProperties);
+    queryErrors
+        = new QueryMetadataImpl.TimeBoundedQueue(Duration.ofHours(1), maxQueryErrorsQueueSize);
+    this.streamsProperties = ImmutableMap.copyOf(streamsProperties);
+    metadata = new ConcurrentHashMap<>();
+    sources = new ConcurrentHashMap<>();
+    kafkaStreams.setUncaughtExceptionHandler(this::uncaughtHandler);
+    kafkaStreams.start();
+  }
+
+  public void markSources(final QueryId queryId, final Set<SourceName> sourceNames) {
+    sources.put(queryId, sourceNames);
+  }
+
+  public void addQuery(
+          final QueryErrorClassifier errorClassifier,
+          final Map<String, Object> streamsProperties,
+          final PersistentQueriesInSharedRuntimesImpl persistentQueriesInSharedRuntimesImpl,
+          final QueryId queryId) {
+    this.errorClassifier = errorClassifier;
+    if (metadata.containsKey(queryId.toString())) {
+      kafkaStreams.removeNamedTopology(queryId.toString());
```

Comment:
When this scenario can actually happen? And should we really just remove the old query silently?

Reply:
There is a possible race where is the query was added and then replaced the `stop()` might not be able to remove the topology, once the `removeNamedTopology` returns a `future` we can remove this check. we swallow the possible error is stop to clean up the state and let the other queries continue processing so we might need to remove it again.

Reply:
Okay that makes sense. For the second part `we swallow the possible error is stop to clean up the state and let the other queries continue processing so we might need to remove it again.` though, not sure I fully follow that (maybe it's just due to typos..) could you re-explain to me?

Reply:
wow I butchered that sentence.... I just remove the swallowing of the error so it shouldn't be an issue anymore